### ISSUE TITLE
locks one cpu core at 100%

### ISSUE THREAD

**Issue by jcanderle** [mlichvar/timedatex#4]
I had to disable the systemd service, as this timedatex is keeping one core at 100%
Fedora 26 4.11.0-2.fc26.x86_64 #1 SMP Tue May 9 15:24:49 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**Comment by mlichvar** [comment]
It is a bug in glib2. It was already reported as issue #3. There is a glib2 update in F26 which should fix it: 
https://bodhi.fedoraproject.org/updates/FEDORA-2017-27fbf57af3

I'll leave this open for a while in case more people are searching for the issue here.

**Comment by jcanderle** [comment]
Thanks, upgrading to glib2-2.52.2-2.fc26.x86_64.rpm fixed the issue.

**Comment by kalev** [comment]
This regressed again with glib 2.54.3 where the upstream patch was reverted; see https://bugzilla.gnome.org/show_bug.cgi?id=761102#c52 for details.

I discussed this with the upstream people and they would appreciate help with debugging this, especially from people familiar with timedatex code.

**Comment by pwithnall** [comment]
Looks like you’re missing a `g_main_context_wakeup()` call in `stop_main_loop()` after changing the termination condition.

**Comment by pwithnall** [comment]
(That’s just what I can spot from quickly looking at the timedatex code. I don’t know if it will fix the issue. If you’re still getting the 100% CPU issue with that change, we’d need a backtrace of the infinite loop to debug it.)

**Comment by mlichvar** [comment]
Adding `g_main_context_wakeup()` call to `stop_main_loop()` doesn't seem to fix the issue. (It's not  very clear to me why this is needed.)

The problem is the same as before. `g_main_context_iteration()` returns immediately if the timeout has been added. Here is a minimal reproducer:

```
#include <glib.h>
#include <glib/gstdio.h>

static gboolean main_quit = FALSE;

static gboolean stop_main_loop(gpointer user_data) {
        main_quit = TRUE;
        return TRUE;
}

int main(int argc, char **argv) {
        guint timeout_id = 0;
        int iter = 0;

        while (!main_quit) {
                timeout_id = g_timeout_add(30 * 1000, stop_main_loop, NULL);

                fprintf(stderr, "iteration %d\n", iter++);
                g_main_context_iteration(g_main_context_default(), TRUE);

                if (timeout_id)
                        g_source_remove(timeout_id);
                timeout_id = 0;
        }

        return 0;
}
```

**Comment by mlichvar** [comment]
I should add that `g_main_context_iteration()` returns FALSE. According to the documentation that is possible:

> Note that even when may_block is TRUE, it is still possible for g_main_context_iteration() to return FALSE, since the wait may be interrupted for other reasons than an event source becoming ready.

The question is if "other reasons" may include a newly added timeout.

When I modify the code to not remove the timeout if the function returns FALSE, it seems to work as expected.

**Comment by pwithnall** [comment]
> The question is if "other reasons" may include a newly added timeout.

Quite possibly, and that is precisely the code which the [reverted commit](https://git.gnome.org/browse/glib/commit/?id=9ba95e25b74adf8d62effeaf6567074ac932811c) in GLib touches. With or without that commit, you should be prepared for `g_main_context_iteration()` to return `FALSE`.

Instead of installing a timeout callback on each main context iteration, I suggest you install it once, before the first main context iteration, and in the callback you check a `last_activity_time` variable against your timeout duration to work out whether to quit the main loop. You can then set `last_activity_time = g_get_monotonic_time()` after each event you want to reset the timeout. (Alternatively, you can add/reset the timeout callback after each such event.)

Doing your timeout handling on each `g_main_context_iteration()` call means it will be spuriously reset by unrelated events, such as for unrelated D-Bus traffic (signal broadcasts from other services) arriving on the D-Bus FD and then being filtered in software. Doing the timeout handling here is too low-level.

**Comment by mlichvar** [comment]
Thanks for the suggestions. It should be now fixed with commit 3f0b85db1d4eedd3fdc9db0832e114215812d15c.
